File: Packages/HistoryKit/Tests/HistoryKitTests/RestoreStackTests.swift
import TestKit
import Foundation
@testable import HistoryKit

let restoreStackTests: [TestCase] = [
    TestCase("popReturnsNewestFirst") { t in
        var s = RestoreStack()
        let a = UUID(), b = UUID()
        s.push(a); s.push(b)
        t.equal(s.pop(), b)
        t.equal(s.pop(), a)
        t.isNil(s.pop())
    },
    TestCase("isEmptyTracksContents") { t in
        var s = RestoreStack()
        t.isTrue(s.isEmpty)
        s.push(UUID())
        t.isFalse(s.isEmpty)
        _ = s.pop()
        t.isTrue(s.isEmpty)
    },
    TestCase("depthCapDropsOldest") { t in
        var s = RestoreStack()
        let ids = (0..<6).map { _ in UUID() }
        for id in ids { s.push(id) }
        // ids[0] fell off the bottom; newest five remain in LIFO order.
        for id in ids.dropFirst().reversed() { t.equal(s.pop(), id) }
        t.isNil(s.pop())
    },
    TestCase("repushMovesIDToTop") { t in
        var s = RestoreStack()
        let a = UUID(), b = UUID()
        s.push(a); s.push(b); s.push(a)
        t.equal(s.pop(), a)
        t.equal(s.pop(), b)
        t.isNil(s.pop())
    },
]

File: windows/tests/BetterScreenshot.Tests/RestoreStackTests.cs
using BetterScreenshot.History;
using Xunit;

namespace BetterScreenshot.Tests;

public class RestoreStackTests
{
    [Fact]
    public void PopReturnsNewestFirst()
    {
        var s = new RestoreStack();
        var a = Guid.NewGuid();
        var b = Guid.NewGuid();
        s.Push(a);
        s.Push(b);
        Assert.Equal(b, s.Pop());
        Assert.Equal(a, s.Pop());
        Assert.Null(s.Pop());
    }

    [Fact]
    public void IsEmptyTracksContents()
    {
        var s = new RestoreStack();
        Assert.True(s.IsEmpty);
        s.Push(Guid.NewGuid());
        Assert.False(s.IsEmpty);
        s.Pop();
        Assert.True(s.IsEmpty);
    }

    [Fact]
    public void DepthCapDropsOldest()
    {
        var s = new RestoreStack();
        var ids = new List<Guid>();
        for (int k = 0; k < RestoreStack.Depth + 1; k++)
        {
            var id = Guid.NewGuid();
            ids.Add(id);
            s.Push(id);
        }
        // Newest 5 remain, popped newest-first; the very first pushed id was dropped.
        for (int k = ids.Count - 1; k >= 1; k--)
            Assert.Equal(ids[k], s.Pop());
        Assert.Null(s.Pop()); // ids[0] was dropped
    }

    [Fact]
    public void RepushMovesIdToTop()
    {
        var s = new RestoreStack();
        var a = Guid.NewGuid();
        var b = Guid.NewGuid();
        s.Push(a);
        s.Push(b);
        s.Push(a); // move a to top
        Assert.Equal(a, s.Pop());
        Assert.Equal(b, s.Pop());
    }
}
